File: algorithms/1572.matrix-diagonal-sum/solution.rb
# @param {Integer[][]} mat
# @return {Integer}
def diagonal_sum(mat)
  sum = 0
  i = mat.size / 2

  if mat.size % 2 == 1
    sum += mat[i][i]
  end

  i -= 1

  while i >= 0
    sum += mat[i][i] + mat[i][-i-1] + mat[-i-1][i] + mat[-i-1][-i-1]
    i -= 1
  end

  sum
end

File: algorithms/1572.matrix-diagonal-sum/solution.rs
impl Solution {
    pub fn diagonal_sum(mat: Vec<Vec<i32>>) -> i32 {
        let mut sum: i32 = 0;
        let len = mat.len();
        let mut i: usize = len / 2;

        if mat.len() % 2 == 1 {
            sum += mat[i][i];
        }

        while i > 0 {
            i -= 1;
            sum += mat[i][i]
                + mat[i][len - i - 1]
                + mat[len - i - 1][i]
                + mat[len - i - 1][len - i - 1];
        }

        sum
    }
}
